internals/daemon/api_notices_test.go

```diff
@@ -205,19 +251,202 @@ func (s *apiSuite) TestNoticesFilterInvalidTypes(c *C) {
+		userIDValues.Add("user-id", strconv.FormatUint(uint64(uid), 10))
+		reqUrl := fmt.Sprintf("/v1/notices?%s", userIDValues.Encode())
+		req, err := http.NewRequest("GET", reqUrl, nil)
+		c.Assert(err, IsNil)
+		req.RemoteAddr = "pid=100;uid=0;socket=;"
+		rsp, ok := noticesCmd.GET(noticesCmd, req, nil).(*resp)
+		c.Assert(ok, Equals, true)
+
+		c.Check(rsp.Type, Equals, ResponseTypeSync)
+		c.Check(rsp.Status, Equals, http.StatusOK)
+		notices, ok := rsp.Result.([]*state.Notice)
+		c.Assert(ok, Equals, true)
+		c.Assert(notices, HasLen, 2)
+		n := noticeToMap(c, notices[0])
+		c.Assert(n["user-id"], Equals, float64(uid))
+		n = noticeToMap(c, notices[1])
+		c.Assert(n["user-id"], Equals, nil)
+	}
+}
+
+func (s *apiSuite) TestNoticesUserIDsNonAdminDefault(c *C) {
+	s.daemon(c)
+	restore := fakeSysGetuid(0)
+	defer restore()
+
+	st := s.d.overlord.State()
+	st.Lock()
+	admin := uint32(0)
+	nonAdmin := uint32(1000)
+	otherNonAdmin := uint32(123)
+	addNotice(c, st, &admin, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, &nonAdmin, state.CustomNotice, "a.b/x", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, &otherNonAdmin, state.CustomNotice, "a.b/y", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, nil, state.WarningNotice, "danger", nil)
+	st.Unlock()
+
+	noticesCmd := apiCmd("/v1/notices")
+
+	// Test that non-admin user by default only sees their notices and public notices.
+	req, err := http.NewRequest("GET", "/v1/notices", nil)
+	c.Assert(err, IsNil)
+	req.RemoteAddr = "pid=100;uid=1000;socket=;"
+	rsp, ok := noticesCmd.GET(noticesCmd, req, nil).(*resp)
+	c.Assert(ok, Equals, true)
+
+	c.Check(rsp.Type, Equals, ResponseTypeSync)
+	c.Check(rsp.Status, Equals, http.StatusOK)
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 2)
+	n := noticeToMap(c, notices[0])
+	c.Assert(n["user-id"], Equals, float64(nonAdmin))
+	c.Assert(n["key"], Equals, "a.b/x")
+	n = noticeToMap(c, notices[1])
+	c.Assert(n["user-id"], Equals, nil)
+	c.Assert(n["key"], Equals, "danger")
+}
+
+func (s *apiSuite) TestNoticesUserIDsNonAdminFilter(c *C) {
+	s.daemon(c)
+	restore := fakeSysGetuid(0)
+	defer restore()
+
+	st := s.d.overlord.State()
+	st.Lock()
+	nonAdmin := uint32(1000)
+	addNotice(c, st, &nonAdmin, state.CustomNotice, "a.b/x", nil)
+	st.Unlock()
+
+	noticesCmd := apiCmd("/v1/notices")
+
+	// Test that non-admin user may not use --user-id filter
+	reqUrl := "/v1/notices?user-id=1000"
+	req, err := http.NewRequest("GET", reqUrl, nil)
+	c.Assert(err, IsNil)
+	req.RemoteAddr = "pid=100;uid=1000;socket=;"
+	rsp, ok := noticesCmd.GET(noticesCmd, req, nil).(*resp)
+	c.Assert(ok, Equals, true)
+
+	c.Check(rsp.Type, Equals, ResponseTypeError)
+	c.Check(rsp.Status, Equals, http.StatusForbidden)
+	result, ok := rsp.Result.(*errorResult)
+	c.Assert(ok, Equals, true)
+	c.Assert(result.Message, Matches, `only admins may use the "user-id" filter`)
```

Comment:
I'd say we probably don't want to test the text of this Forbidden error message here, just that the status is Forbidden above is probably enough. (I don't think we want this error text to be part of the contract.) Same throughout.